File: abeln-hb-lattice-2e567b22ecd1/source/ClusterMove.cpp
#define MAX_TRANSLATION 10


int ClusterMove(){
  /// calculate clusters
  Cluster cl;
  cl.startCounting(mainLattice);
  int numClusts = cl.getTotalClusters();

  /// pick random clusters
  int cluster2move = (int) floor(numClusts * drand48());

  /// get all chains in cluster
  vector<int> chainList = cl.getChains(cluster2move);

  /// set rotation and translation
  //int dir_translation= (int)floor(6 * drand48());
  //int magnitude_translation = (int)floor(MAX_TRANSLATION * drand48());
  int rotationalDir= (int)floor(6 * drand48());

  /// choose random pivot on lattice
  int xpivot =  (int)floor(LX * drand48()); 
  int ypivot =  (int)floor(LY * drand48());
  int zpivot =  (int)floor(LZ * drand48());
  Pos posPivot =  Pos(xpivot,ypivot,zpivot);

  vector< vector Pos>> newPos;
  
  /// for each chain, perform rotation and translation,
  /// while keeping old coordnates
  /// check for clashes or touches: reject, restore old coordinates and sample
  for(int i=0; i< chainList.size();i++){
    int cn = chainList[i];
    Chain * ch = l->chains[cn];    
    for(int n=ch->startRes; n<=ch->endRes;n++){
      Residue * res = ch->residues[n];
      posNew[cn][n]=  ch->rotationPosition(res->pos,rotationalDir,posPivot);
      if(l->checkForClashesAndTouches(posNew[cn][n])){
	/// reject move
	l->sampleCurrentStats();
	return -1;
      }      

    }
  }
  /// now we know there are no clashes our touches
  /// and the move can be performed
  /// note that the move will alsways get accepted, since dE=0


  /// Perform move, update lattice
  for(int i=0; i< chainList.size();i++){
    int cn = chainList[i];
    Chain * ch = l->chains[cn];
    for(int n=ch->startRes; n<=ch->endRes;n++){
      Residue * res = ch->residues[n];
      Pos newSpin = ch->rotationSpin(res->spin,rotationalDir);
      l->emptyLatticePos(res->pos);

      res->pos = posNew[cn][n];
      res->spin = newSpin;
      l->setResidue(posNew[cn][n],res)
    }
  }
}

File: nelman-kmcLATTICEMODEL/source/ClusterMove.py
import random
from typing import List, Tuple
from Cluster import Cluster
from Lattice import Lattice
from Chain import Chain
from Residue import Residue
from Pos import Pos

MAX_TRANSLATION = 10

def cluster_move(main_lattice: Lattice) -> int:
    """ Perform a cluster move on the lattice."""
    cl = Cluster()
    cl.startCounting(main_lattice)
    num_clusters = cl.getTotalClusters()
    
    if num_clusters == 0:
        return -1
        
    # Pick random cluster
    cluster_to_move = random.randint(0, num_clusters - 1)
    chain_list = cl.getChains(cluster_to_move)
    
    # Set rotation
    rotational_dir = random.randint(0, 5)
    
    # Choose random pivot
    x_pivot = random.randint(0, main_lattice.LX - 1)
    y_pivot = random.randint(0, main_lattice.LY - 1)
    z_pivot = random.randint(0, main_lattice.LZ - 1)
    pos_pivot = Pos(x_pivot, y_pivot, z_pivot)
    
    # Store new positions
    new_positions = []
    
    # Check for clashes
    for i, cn in enumerate(chain_list):
        ch = main_lattice.chains[cn]
        chain_positions = []
        for n in range(ch.N):
            res = ch.residues[n]
            new_pos = ch.rotationPosition(res.pos, rotational_dir, pos_pivot)
            chain_positions.append(new_pos)
            if main_lattice.checkForClashesAndTouches(new_pos):
                return -1
        new_positions.append(chain_positions)
    
    # Perform the move
    for i, cn in enumerate(chain_list):
        ch = main_lattice.chains[cn]
        for n in range(ch.N):
            res = ch.residues[n]
            new_spin = ch.rotationSpin(res.spin, rotational_dir)
            main_lattice.emptyLatticePos(res.pos)
            
            res.pos = new_positions[i][n]
            res.spin = new_spin
            main_lattice.setResidue(new_positions[i][n], res)
    
    return 1
